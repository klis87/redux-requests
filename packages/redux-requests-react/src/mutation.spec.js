import renderer from 'react-test-renderer';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Mutation from './mutation';

const mockStore = configureStore();

describe('Mutation', () => {
  const MUTATION_TYPE = 'MUTATION_TYPE';

  it('supports custom selector', () => {
    const component = renderer.create(
      <Provider
        store={mockStore({
          request: {
            mutations: {
              [MUTATION_TYPE]: { pending: 1, error: 'error', ref: {} },
            },
          },
        })}
      >
        <Mutation selector={state => state.request.mutations[MUTATION_TYPE]}>
          {({ loading, error }) => (
            <div>
              {loading && 'loading'}
              {error}
            </div>
          )}
        </Mutation>
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('maps type to mutation', () => {
    const component = renderer.create(
      <Provider
        store={mockStore({
          requests: {
            mutations: {
              [MUTATION_TYPE]: { pending: 1, error: 'error', ref: {} },
            },
          },
        })}
      >
        <Mutation type={MUTATION_TYPE}>
          {({ loading, error }) => (
            <div>
              {loading && 'loading'}
              {error}
            </div>
          )}
        </Mutation>
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('maps type to default mutation when no type found', () => {
    const component = renderer.create(
      <Provider
        store={mockStore({
          requests: {
            mutations: {},
          },
        })}
      >
        <Mutation type={MUTATION_TYPE}>
          {({ loading, error }) => (
            <div>
              {loading && 'loading'}
              {error}
            </div>
          )}
        </Mutation>
      </Provider>,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders loading and error of mutation with requestKey', () => {
    const component = renderer.create(
      <Provider
        store={mockStore({
          requests: {
            mutations: {
              [`${MUTATION_TYPE}x`]: { pending: 1, error: 'error', ref: {} },
            },
          },
        })}
      >
        <Mutation type={MUTATION_TYPE} requestKey="x">
          {({ loading, error }) => (
            <div>
              {loading && 'loading'}
              {error}
            </div>
          )}
        </Mutation>
      </Provider>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('doesnt crush if an mutation with a requestKey doesnt exist', () => {
    const component = renderer.create(
      <Provider
        store={mockStore({
          requests: {
            mutations: {
              [MUTATION_TYPE]: { x: { pending: 1, error: 'error', ref: {} } },
            },
          },
        })}
      >
        <Mutation type={MUTATION_TYPE} requestKey="y">
          {({ loading, error }) => (
            <div>
              {loading && 'loading'}
              {error}
            </div>
          )}
        </Mutation>
      </Provider>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders custom component as prop with extra prop', () => {
    const MutationComponent = ({ mutation: { loading, error }, extra }) => (
      <div>
        {loading && 'loading'} {error} {extra}
      </div>
    );

    const component = renderer.create(
      <Provider
        store={mockStore({
          requests: {
            mutations: {
              [MUTATION_TYPE]: { pending: 1, error: 'error', ref: {} },
            },
          },
        })}
      >
        <Mutation
          type={MUTATION_TYPE}
          extra="extra"
          component={MutationComponent}
        />
      </Provider>,
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
