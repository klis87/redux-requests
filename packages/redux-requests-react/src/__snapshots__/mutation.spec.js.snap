// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation doesnt crush if an mutation with a requestKey doesnt exist 1`] = `<div />`;

exports[`Mutation maps type to default mutation when no type found 1`] = `<div />`;

exports[`Mutation maps type to mutation 1`] = `
<div>
  loading
  error
</div>
`;

exports[`Mutation renders custom component as prop with extra prop 1`] = `
<div>
  loading
   
  error
   
  extra
</div>
`;

exports[`Mutation renders loading and error of mutation with requestKey 1`] = `
<div>
  loading
  error
</div>
`;

exports[`Mutation supports custom selector 1`] = `
<div>
  error
</div>
`;
